utils/upload.go

```diff
@@ -0,0 +1,171 @@
+package utils
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"net/url"
+	"os"
+	"path/filepath"
+	"zestream-server/constants"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+
+	"cloud.google.com/go/storage"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+)
+
+// channel to extract files from the folder
+type fileWalk chan string
+
+type Uploader interface {
+	Upload(walker fileWalk)
+}
+
+func UploadToCloudStorage(uploader Uploader, path string) {
+	walker := make(fileWalk)
+	go func() {
+		//get files to upload via the channel
+		if err := filepath.Walk(path, walker.WalkFunc); err != nil {
+			log.Fatalln("Walk failed: ", err)
+		}
+
+		close(walker)
+
+	}()
+
+	uploader.Upload(walker)
+
+}
+
+func (f fileWalk) WalkFunc(path string, info os.FileInfo, err error) error {
+	if err != nil {
+		return err
+	}
+
+	if !info.IsDir() {
+		f <- path
+	}
+
+	return nil
+}
+
+type AwsUploader struct {
+	Prefix string
+	//common session to be used by every upload
+	Session *session.Session
+}
+
+func (a AwsUploader) Upload(walker fileWalk) {
+	bucket := constants.S3_BUCKET_NAME
+	if bucket == "" {
+		log.Fatalln("AWS Bucketname not available")
+	}
+	log.Printf("bucket %s", bucket)
+
+	prefix := a.Prefix
+
+	uploader := s3manager.NewUploader(a.Session)
+	for path := range walker {
+		filename := filepath.Base(path)
+
+		file, err := os.Open(path)
```

Comment:
Here, if `os.Open(path)` returns an error still the execution continues, in my opinion it should return an error upon a certain number of re-try. What do you think about it??

We can implement the re-try logic later on as well. Same goes in `GcpUploader` and `AzureUploader`

Reply:
Right. I used continue because i did not want the function to stop upload of other files when `os.Open(path)` returns an error. I did print the error message with file path. Should we stop the function when an error occurs?

Reply:
Yeah seems logical to skip a file if we encounter error while opening it, Thanks for the explanation!!